AccessControlException from CipherCore when Security manager enabled

The following test cases are failed with openssl:
com/sun/crypto/provider/Cipher/AES/TestAESCiphers/TestAESWithProviderChange.java.TestAESWithProviderChange
com/sun/crypto/provider/Cipher/AES/TestAESCiphers/TestAESWithRemoveAddProvider.java.TestAESWithRemoveAddProvider

https://hyc-runtimes-jenkins.swg-devops.com/view/Test_security/job/openjdk8_j9_jdksecurity1test_x86-64_linux/18/

Caused by: java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "accessClassInPackage.sun.security.action")
	at java.security.AccessController.throwACE(AccessController.java:176)
	at java.security.AccessController.checkPermissionHelper(AccessController.java:237)
	at java.security.AccessController.checkPermission(AccessController.java:373)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
	at java.lang.SecurityManager.checkPackageAccess(SecurityManager.java:1564)
	at java.lang.J9VMInternals$2.run(J9VMInternals.java:268)
	at java.security.AccessController.doPrivileged(AccessController.java:673)
	at java.lang.J9VMInternals.checkPackageAccess(J9VMInternals.java:263)
	at com.sun.crypto.provider.CipherCore.<clinit>(CipherCore.java:1231)

The issue is that from com.sun.crypto.provider.CipherCore, we are trying to access the class sun.security.action.GetPropertyAction. CipherCore is loaded with extension classloader and GetPropertyAction is loaded with boot class loader. This will result in a security exception as package.access restriction is defined in java.security. We can workaround by providing the RuntimePermission in the policy file. However, that is not a good approach as it is against the definition in java.security.. We need to read the property differently to avoid the issue..

Also, it makes more sense to have the value of jdk.nativeCrypto as 'yes' or 'no' instead of 'true' or 'false' as we need to change the default to enable native crypto when code stabilizes and then jdk.nativeCrypto=no instead of jdk.nativeCrypto=false to disable native crypto makes more sense.

--- a/jdk/src/share/classes/com/sun/crypto/provider/CipherCore.java
+++ b/jdk/src/share/classes/com/sun/crypto/provider/CipherCore.java
@@ -39,8 +39,9 @@ import java.security.spec.*;
 import javax.crypto.*;
 import javax.crypto.spec.*;
 import javax.crypto.BadPaddingException;
-import sun.security.action.GetPropertyAction;
 import jdk.crypto.jniprovider.NativeCrypto;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 
 /**
  * This class represents the symmetric algorithms in its various modes
@@ -65,19 +66,19 @@ final class CipherCore {
     /*
      * Check whether native crypto is enabled with property.
      *
-     * By default, the native crypto is disabled and use java crypto
-     * implementation.
+     * By default, the native crypto is enabled and uses the native 
+     * crypto library implementation.
      *
      * The property 'jdk.nativeCBC' is used to enable Native CBC alone,
      * 'jdk.nativeGCM' is used to enable Native GCM alone and
      * 'jdk.nativeCrypto' is used to enable all native cryptos (Digest,
      * CBC and GCM).
      */
-    private static boolean useNativeCrypto = false;
+    private static boolean useNativeCrypto = true;
 
-    private static boolean useNativeCBC = false;
+    private static boolean useNativeCBC = true;
 
-    private static boolean useNativeGCM = false;
+    private static boolean useNativeGCM = true;
 
     /*
      * internal buffer
@@ -1226,18 +1227,28 @@ final class CipherCore {
         cipher.updateAAD(src, offset, len);
     }
 
-    static {
-        useNativeCrypto = Boolean.parseBoolean(
-                GetPropertyAction.privilegedGetProperty("jdk.nativeCrypto"));
+    private static String privilegedGetProperty(final String property) {
+        return AccessController.doPrivileged(new PrivilegedAction<String>() {
+            public String run() {
+                return System.getProperty(property);
+            }
+        });
+    }
 
-        if (useNativeCrypto) {
-            useNativeCBC = true;
-            useNativeGCM = true;
+    static {
+        String nativeCryptTrace = privilegedGetProperty("jdk.nativeCryptoTrace");
+        String nativeCryptStr   = privilegedGetProperty("jdk.nativeCrypto");
+        String nativeCBCStr     = privilegedGetProperty("jdk.nativeCBC");
+        String nativeGCMStr     = privilegedGetProperty("jdk.nativeGCM");
+
+        useNativeCrypto = Boolean.parseBoolean(nativeCryptStr) || (nativeCryptStr == null);
+
+        if (!useNativeCrypto) {
+            useNativeCBC = false;
+            useNativeGCM = false;
         } else {
-            useNativeCBC = Boolean.parseBoolean(
-                    GetPropertyAction.privilegedGetProperty("jdk.nativeCBC"));
-            useNativeGCM = Boolean.parseBoolean(
-                    GetPropertyAction.privilegedGetProperty("jdk.nativeGCM"));
+            useNativeCBC = Boolean.parseBoolean(nativeCBCStr) || (nativeCBCStr == null);
+            useNativeGCM = Boolean.parseBoolean(nativeGCMStr) || (nativeGCMStr == null);
         }
 
         if (useNativeCBC || useNativeGCM) {
@@ -1248,11 +1259,22 @@ final class CipherCore {
              * java crypto implementation.
              */
             if (!NativeCrypto.isLoaded()) {
-                System.err.println("Warning: Native crypto library load failed." +
-                                   " Using Java crypto implementation");
                 useNativeCBC = false;
                 useNativeGCM = false;
+
+		if (nativeCryptTrace != null) {
+                   System.err.println("Warning: Native crypto library load failed." +
+                                   " Using Java crypto implementation");
+		}
+            } else {
+                if (nativeCryptTrace != null) {
+                   System.err.println("CipherCore Load - using native crypto library.");
+                }
             }
-        }
+        } else {
+                if (nativeCryptTrace != null) {
+                   System.err.println("CipherCore Load - native crypto library disabled.");
+                }
+	}
     }
 }

--- a/jdk/src/share/classes/sun/security/provider/SunEntries.java
+++ b/jdk/src/share/classes/sun/security/provider/SunEntries.java
@@ -92,12 +92,12 @@ final class SunEntries {
 
     /*
      * Check whether native crypto is enabled with property.
-     * By default, the native crypto is disabled and use sun crypto providers.
+     * By default, the native crypto is enabled and uses native library crypto.
      * The property 'jdk.nativeDigest' is used to enable Native digest alone
      * and 'jdk.nativeCrypto' is used to enable all native cryptos (Digest,
      * CBC and GCM).
      */
-    private static boolean useNativeDigest = false;
+    private static boolean useNativeDigest = true;
 
     private SunEntries() {
         // empty
@@ -420,10 +420,20 @@ final class SunEntries {
     }
 
     static {
-        useNativeDigest = Boolean.parseBoolean(
-                              GetPropertyAction.privilegedGetProperty("jdk.nativeCrypto")) ||
-                          Boolean.parseBoolean(
-                              GetPropertyAction.privilegedGetProperty ("jdk.nativeDigest"));
+
+        String nativeCryptTrace = GetPropertyAction.privilegedGetProperty("jdk.nativeCryptoTrace");
+        String nativeCryptStr = GetPropertyAction.privilegedGetProperty("jdk.nativeCrypto");
+        String nativeDigestStr = GetPropertyAction.privilegedGetProperty("jdk.nativeDigest");
+
+        if (Boolean.parseBoolean(nativeCryptStr) || nativeCryptStr == null) {
+                /* nativeCrypto is enabled */
+                if (!(Boolean.parseBoolean(nativeDigestStr) || nativeDigestStr == null)) {
+                        useNativeDigest = false;
+                }
+        } else {
+                /* nativeCrypto is disabled */
+                useNativeDigest = false;
+        }
 
         if (useNativeDigest) {
             /*
@@ -433,9 +443,20 @@ final class SunEntries {
              * java crypto implementation.
              */
             if (!NativeCrypto.isLoaded()) {
-                System.err.println("Warning: Native crypto library load failed." +
-                                   " Using Java crypto implementation");
                 useNativeDigest = false;
+
+                if (nativeCryptTrace != null) {
+                   System.err.println("Warning: Native crypto library load failed." +
+                                   " Using Java crypto implementation");
+                }
+            } else {
+                if (nativeCryptTrace != null) {
+                   System.err.println("MessageDigest load - using Native crypto library.");
+                }
+            }
+        } else {
+            if (nativeCryptTrace != null) {
+               System.err.println("MessageDigest load - Native crypto library disabled.");
             }
         }
     }

